cohttp-lwt/src/s.ml

```diff
@@ -106,13 +106,27 @@ module type Server = sig
 
   type conn = IO.conn * Cohttp.Connection.t
 
+  type response_action =
+    (* The connection is not closed in the [`Expert] case until the [unit
+       Lwt.t] becomes determined. *)
+    [ `Expert of Cohttp.Response.t
+                 * (IO.ic
+                    -> IO.oc
+                    -> unit Lwt.t)
+    | `Response of Cohttp.Response.t * Body.t ]
+
   type t
 
   val make : ?conn_closed:(conn -> unit)
     -> callback:(conn -> Cohttp.Request.t -> Body.t
                  -> (Cohttp.Response.t * Body.t) Lwt.t)
     -> unit -> t
 
+  val make_expert : ?conn_closed:(conn -> unit)
```

Comment:
I'm wondering if this shouldn't eventually become the default function, so it shouldn't be called `make_expert`.  I suggest three functions:

`make_raw` which returns what `make_expert` does above
`make_expert` which wraps the response of `make_raw` in a `Expert` value
`make` which wraps the response of `make_raw` in a `Response` value

Reply:
Is this what you mean?

```ocaml
val make_raw :
     ?conn_closed:(conn -> unit)
  -> callback:(conn -> Cohttp.Request.t -> Body.t -> response_action Lwt.t)
  -> unit
  -> t

val make_expert :
     ?conn_closed:(conn -> unit)
  -> callback:(   conn
               -> Cohttp.Request.t
               -> Body.t
               -> Cohttp.Response.t * (IO.ic -> IO.oc -> unit Lwt.t) Lwt.t)
  -> unit
  -> t

val make :
     ?conn_closed:(conn -> unit)
  -> callback:(   conn
               -> Cohttp.Request.t
               -> Body.t
               -> (Cohttp.Response.t * Body.t) Lwt.t)
  -> unit
  -> t
```

Reply:
yep! Perhaps `make_raw` would be clearer as `make_response_action`?

Reply:
Updated.